### ISSUE TITLE
Resources usage is crashing HA/RPi

### ISSUE THREAD

**Issue by OToussaint** [mudape/iphonedetect#75]
It is a long time that I'm trying to identify which of my integration is crashing HA first and my RPi next.
And I think I've found it.

The number of opened files was increasing up to a number where no socket connection and/or any other things were possible until restarting HA and its integrations (or worst case scenario, the RPi itself).

I graphed the "lsof" output for 2 monthes, disabling one integration at a time to try to narrow to the one causing my issue and here is my latest result:
![image](https://user-images.githubusercontent.com/58084734/170685511-f7aa33fd-4c0a-4195-a17f-d1ad9540eaf9.png)

In the first part, HA is using the latest version of iphonedetect (v2.0.3)
In the second part, the integration is disabled
In the third part, it is the version 1.0.4 (oldest I can install with HACS)

As you can see, the resources used are dramatically increasing in v2.0.3, not at all without iphonedetect and at a slower pace in v1.0.4

**Comment by mudape** [comment]
Hi @OToussaint,
Thanks for reporting!

Looks like sockets aren't being closed as they should.
Could you explain have you got the numbers for lsof as when I try I'm nowhere near the numbers in the graph.

I used a sensor within HASS
```YAML
sensor:
  - platform: command_line
    name: lsof
    command: 'lsof | wc -l'
```

I track five devices on a /24 network

**Comment by mudape** [comment]
Added context managers for socket and IPRoute in 2.0.4 beta.
On my systems 'lsof' now has stopped increasing.

Please could you try it?
Thanks!

**Comment by OToussaint** [comment]
Hello @mudape, 

So do I for lsof. But I noticed that running the command with the homeassistant user was not reporting everything related to hass.
So I did it a bit differently, I have a cron entry, doing

`*/1 * * * * lsof | wc -l > /home/homeassistant/lsof.tmp; mv /home/homeassistant/lsof.tmp /home/homeassistant/lsof.txt`

And the sensor is
```
  - platform: command_line
    name: lsof
    scan_interval: 60
    command: cat /home/homeassistant/lsof.txt
    value_template: "{{ value | round(0) }}"
```
I have a post in community with a bit more detail if you want. I'll be pleased to help you find the root cause.
I'll install the 2.0.4 beta now and let you know.

**Comment by OToussaint** [comment]
Here is the [link to the community post](https://community.home-assistant.io/t/error-doing-job-socket-accept-out-of-system-resource/413938)
The version 2.0.4 is not solving the issue. Here is the graph after installing it and re-enabling the integration

![image](https://user-images.githubusercontent.com/58084734/170859831-13f880a0-f5ae-4fa3-83b9-b2e2e4578174.png)

**Comment by mudape** [comment]
Thanks for testing!

I've not yet added your cron job, 
here my graphs from the command_line 'lsof | wc -l' sensor
![LSOF Docker Core](https://user-images.githubusercontent.com/26150341/170862314-dfca45df-5499-4e6c-abec-fcd416c61d03.png)
![LSOF RPi4 HASSOS](https://user-images.githubusercontent.com/26150341/170862370-cfeaebba-fcf2-4ef0-a4fd-ff72ec075fe2.png)

I'm guessing HASSOS or other than Core is causing the issue?

**Comment by OToussaint** [comment]
I'm not using HASSOS, I'm using Core.
Maybe it is something else, I don't know, 
But it is completely still when disabling iphonedetect, maybe a bad interraction with something else, indeed.
If you look at my community post, you'll see that most of opened handles are "pipe" wich is not helping as you don't know who is keeping those pipes open and at which end, is it the creator of the pipe, the listener?

**Comment by mudape** [comment]
I've have never tried running HA in an venv, so have no experience of that.
Guess you have your reason to go that route, makes it kind of hard to know what's causing this behavior though.

That iphonedetect somehow triggers increased/high use of resources feels weird.
Thanks to you 2.0.4 is using context manager and I see no increase in 'lsof' anymore in none venv environments.

Did a quick try to install raspbian on Proxmox but it didn't want to play...
I do have spare RPi's but no memory cards...
Read: I can't reproduce your setup.

Are you running lots of other stuff on your RPi, can you try only running HASS?

**Comment by OToussaint** [comment]
With the latest version, the restart occured after 7 hours and 10 minutes, better than every hour.
What do I run on the RPi ... hmm ... well ... a lot of other things indeed:
 - ESPHome
 - NodeRED
 - PiHole
 - MotionEye
for the main things.

The reason why I'm not using HASSOS is that I feel like I don't have control on what I want and, mainly, I never managed to have it to run on my Pi 3B, it is starting than flow the screen with errors so fast that I can't even read what is wrong (and that with all my SD cards, different size, speed or brand). So my only experience with HASSOS is in a VM.

Moreover, I'm using Home Assistant since version 0.67 (April 2018), lovelace was not even a thing at that time and Hoass.Io was a pre-release 0.1. Core was the only real thing and you can imagine all the efforts I've made throughout the years to get what I have today. The idea of restarting everything is already depressing me.

I'll stop all my stuffs listed above and see.

**Comment by mudape** [comment]
`With the latest version, the restart occured after 7 hours and 10 minutes, better than every hour.`
Progress! :)

Didn't mean to question your setup, sorry if my words sounded like that.
Sounds like you have things running fine until this.
I was simply wondering if other external (to HASS) stuff could be the reason for sockets not closing.

Where you using iphonedetect prior to when the issue first appeared in HASS 2022.4.4 ?
I'm thinking Python or something in your virtual env might be the root of the issue
But that's only based on my unexperienced knowledge with venv...

I usually speak Swedish, your English is more then fine for me to understand and I hope the same goes for you.

**Comment by OToussaint** [comment]
I'm using iphonedetect for very long time (1.0.4 sure, maybe before). Way before 2022.4.4 in any case.
I noticed the leak in 2022.4.4 and not before (even thou it was necessary to restart the Pi from time to time but it was a matter of days, not hours). 
So it could be an issue with what 2022.4.4 was including.
But, as far as I remember, it was not the version requiring python 3.9 (it was earlier this year, maybe February), means I did not upgrade python lately, maybe pip but this has nothing to do with it.

I don't know a lot about venv either. My understanding is : as soon as you create the venv, everything inside is copied from the running OS and frozen in time. Whatever you do outside the venv is no longer impacting the inside of the venv (and vice versa). HA can therefore manager all its python packages independently of everything else running on the machine.

**Comment by OToussaint** [comment]
I'm now extremely confused.
I restarted everything but lsof is stable.
The only difference is that I noticed a cups server that was installed (a long time ago for a test I did) and instead of restarting it, I removed it.
Could it be the one that was misbehaving with iphonedetect? I don't really want to test it by re-installing it.
I'll reboot the RPi, to be sure and let you know for a last status. If it is solved, I'll assume a compatibility issue between those 2, or I'll go back to step "stop everything else than HA" but this time, I'll keep a diary.
